connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java

```diff
@@ -202,6 +204,225 @@ public Object valueFrom(Map<String, Object> map) {
+     * @param updatedSchema  updated struct schema
+     * @param whenFound      function to apply when path is found
+     * @param whenNotFound   function to apply when path is not found
+     * @param whenOther      function to apply on fields not matched by path
+     * @return updated data value
+     */
+    public Struct updateValueFrom(
+        Schema originalSchema,
+        Struct originalValue,
+        Schema updatedSchema,
+        StructValueUpdater whenFound,
+        StructValueUpdater whenNotFound,
+        StructValueUpdater whenOther
+    ) {
+        return updateValue(originalSchema, originalValue, updatedSchema, 0, whenFound, whenNotFound, whenOther);
+    }
+
+    private Struct updateValue(
+        Schema originalSchema,
+        Struct originalValue,
+        Schema updateSchema,
+        int step,
+        StructValueUpdater whenFound,
+        StructValueUpdater whenNotFound,
+        StructValueUpdater whenOther
+    ) {
+        Struct updated = new Struct(updateSchema);
+        boolean found = false;
+        for (Field field : originalSchema.fields()) {
+            if (step < steps.size()) {
+                if (steps.get(step).equals(field.name())) {
+                    found = true;
+                    if (step == lastStepIndex()) {
+                        whenFound.apply(
+                            originalValue,
+                            field,
+                            updated,
+                            updateSchema.field(field.name()),
+                            this
+                        );
+                    } else {
+                        if (field.schema().type() == Schema.Type.STRUCT) {
+                            Struct fieldValue = updateValue(
+                                field.schema(),
+                                originalValue.getStruct(field.name()),
+                                updateSchema.field(field.name()).schema(),
+                                step + 1,
+                                whenFound,
+                                whenNotFound,
+                                whenOther
+                            );
+                            updated.put(field.name(), fieldValue);
+                        } else {
+                            // add back to not found and apply others, as only leaf values are updated
+                            found = false;
+                            whenOther.apply(originalValue, field, updated, null, this);
+                        }
+                    }
+                } else {
+                    whenOther.apply(originalValue, field, updated, null, this);
+                }
+            }
+        }
+        if (!found) {
+            whenNotFound.apply(
+                originalValue,
+                null,
+                updated,
+                updateSchema.field(steps.get(step)),
+                this);
+        }
+        return updated;
+    }
```

Comment:
Just for the sake of the `TimestampConverter` SMT, I don't think we need all this complexity. We could get everything we need without custom interfaces and with fewer arguments by doing something like `public Struct updateStruct(Struct original, Schema updatedSchema, BiFunction<Object, Schema, Object> updateField)` (where the last argument accepts the original field and original schema, and returns the new field).

I've sketched out an implementation but feel free to do your own if you prefer:
```java
public Struct updateStruct(
        Struct struct,
        Schema updatedSchema,
        BiFunction<Object, Schema, Object> update
) {
    return updateStruct(
            struct,
            updatedSchema,
            update,
            steps.get(0),
            steps.subList(1, steps.size())
    );
}

private static Struct updateStruct(
        Struct original,
        Schema updatedSchema,
        BiFunction<Object, Schema, Object> update,
        String currentStep,
        List<String> nextSteps
) {
    if (original == null)
        return null;

    Struct result = new Struct(updatedSchema);
    for (Field field : updatedSchema.fields()) {
        String fieldName = field.name();

        if (fieldName.equals(currentStep)) {
            final Object updatedField;

            // Modify this field
            if (nextSteps.isEmpty()) {
                // This is a leaf node
                Object originalField = original.get(fieldName);
                updatedField = update.apply(
                        originalField,
                        original.schema().field(fieldName).schema()
                );
            } else {
                // We have to go deeper
                Struct originalField = requireStructOrNull(original.get(fieldName), "nested field access");
                updatedField = updateStruct(
                        originalField,
                        field.schema(),
                        update,
                        nextSteps.get(0),
                        nextSteps.subList(1, nextSteps.size())
                );
            }

            result.put(fieldName, updatedField);
        } else {
            // Copy over all other fields from the original to the result
            result.put(fieldName, original.get(fieldName));
        }
    }

    return result;
}
```